### PR TITLE
Added AllocMonitor facility

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

This is a general purpose, extensible memory allocation/deallocation monitor systems. See README.md files for details.

#### PR validation:

Code compiles and run on x86 with new unit tests passing.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
assign core

**Comment by Dr15Jones** [comment]
@dan131riley would it be possible for you to try this on ARM? from compiler explorer the mangled names are different. It would be nice to see if we can make this work on ARM as well.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
I updated the proxy test to test all the `operator new` variations. AFAIK there is no way to force a version of `operator delete`, that seems to be up to the compiler to choose.

**Comment by Dr15Jones** [comment]
please test

**Comment by dan131riley** [comment]
> @dan131riley would it be possible for you to try this on ARM? from compiler explorer the mangled names are different. It would be nice to see if we can make this work on ARM as well.

Doesn't even make it to the name mangling part, looks to get tied up in recursive calls to cmalloc() while trying to hook malloc():

```
(gdb) bt
#0  0x0000ffff87279210 in syscall () from /lib64/libc.so.6
#1  0x0000ffff874a2204 in __cxxabiv1::__cxa_guard_acquire (g=0xffff8793a0b0 <guard variable for calloc::original>) at ../../../../libstdc++-v3/libsupc++/guard.cc:319
#2  0x0000ffff8791b574 in calloc (nitems=1, item_size=32) at /home/dsr/CMSSW_13_3_0_pre2/src/PerfTools/AllocMonitorPreload/src/memory_proxies.cc:45
#3  0x0000ffff87723890 in _dlerror_run () from /lib64/libdl.so.2
#4  0x0000ffff87723278 in dlsym () from /lib64/libdl.so.2
#5  0x0000ffff8791d928 in (anonymous namespace)::get<void* (*)(long unsigned int, long unsigned int) noexcept>(const char *) (iName=0xffff8791e4d0 "calloc") at /home/dsr/CMSSW_13_3_0_pre2/src/PerfTools/AllocMonitorPreload/src/memory_proxies.cc:19
#6  0x0000ffff8791b594 in calloc (nitems=1, item_size=32) at /home/dsr/CMSSW_13_3_0_pre2/src/PerfTools/AllocMonitorPreload/src/memory_proxies.cc:45
#7  0x0000ffff87723890 in _dlerror_run () from /lib64/libdl.so.2
#8  0x0000ffff87723278 in dlsym () from /lib64/libdl.so.2
#9  0x0000ffff8791d8d4 in (anonymous namespace)::get<void* (*)(long unsigned int) noexcept>(const char *) (iName=0xffff8791e4c8 "malloc") at /home/dsr/CMSSW_13_3_0_pre2/src/PerfTools/AllocMonitorPreload/src/memory_proxies.cc:19
#10 0x0000ffff8791b3d4 in malloc (size=72704) at /home/dsr/CMSSW_13_3_0_pre2/src/PerfTools/AllocMonitorPreload/src/memory_proxies.cc:35
#11 0x0000ffff8749dfb4 in (anonymous namespace)::pool::pool (this=0xffff876023b8 <(anonymous namespace)::emergency_pool>) at ../../../../libstdc++-v3/libsupc++/eh_alloc.cc:123
#12 __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at ../../../../libstdc++-v3/libsupc++/eh_alloc.cc:262
#13 _GLOBAL__sub_I_eh_alloc.cc(void) () at ../../../../libstdc++-v3/libsupc++/eh_alloc.cc:338
#14 0x0000ffff8794a22c in call_init.part () from /lib/ld-linux-aarch64.so.1
#15 0x0000ffff8794a330 in _dl_init () from /lib/ld-linux-aarch64.so.1
#16 0x0000ffff8793d46c in _dl_start_user () from /lib/ld-linux-aarch64.so.1
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb)
```

**Comment by Dr15Jones** [comment]
@dan131riley Thanks for giving it a try. IF I found the right version of the code and IF I have interpreted it correctly, it looks like the dlsym is trying to do a one time setup of space for this thread's error reporting. It might be possible to identify that case and handle it by returning a previously _file scope static_ allocated memory block. Would mean `free` would have to watch out for that address as well.

**Comment by dan131riley** [comment]
@Dr15Jones some prior art:

- https://stackoverflow.com/questions/17803456/an-alternative-for-the-deprecated-malloc-hook-functionality-of-glibc
- https://stackoverflow.com/questions/6083337/overriding-malloc-using-the-ld-preload-mechanism
- https://www.slideshare.net/tetsu.koba/tips-of-malloc-free (slide 35)

Generally seems to be having a special memory allocator just for that one allocation.  If we were using the glibc allocator all the time, we could call `__libc_calloc()` instead of getting it from `dlsym()`, but that won't work with (je|tc)malloc.

**Comment by dan131riley** [comment]
@Dr15Jones I hacked together various bits and pieces at [dan131riley:AllocMonitor](https://github.com/dan131riley/cmssw/blob/AllocMonitor/PerfTools/AllocMonitorPreload/src/memory_proxies.cc#L10-L25), which seems to work, or at least passes tests.  I would have been surprised if the mangling were actually different--AFAIK, gcc and llvm follow the[ itanium ABI for mangling](https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling) on all platforms.

**Comment by Dr15Jones** [comment]
@dan131riley Thanks for exploring further! The only issue I have with your implementation is it doesn't guarantee alignment rules if there is more than one allocation request to the local_* functions.

**Comment by dan131riley** [comment]
@Dr15Jones I've updated to fix the alignment issue, good catch.  At least on ARM, there are two allocations--the first one is for an exception handling emergency pool, and the allocation for that then triggers the call to dlsym() that results in the second allocation.

**Comment by Dr15Jones** [comment]
please test

File: PerfTools/AllocMonitor/test/BuildFile.xml
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply PerfTools/AllocMonitor/test/BuildFile.xml:1]: done

File: PerfTools/AllocMonitorPreload/src/memory_proxies.cc [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+    assert(original);
+    return reinterpret_cast<T>(original);
+  }
+
+#ifdef USE_LOCAL_MALLOC
+  // this is a very simple-minded allocator used for any allocations
+  // before we've finished our setup.  In particular, this avoids a
+  // chicken/egg problem if dlsym() allocates any memory.
+  constexpr auto max_align = alignof(std::max_align_t);
+  alignas(max_align) char tmpbuff[131072];
```
Comment (makortel) [inline PerfTools/AllocMonitorPreload/src/memory_proxies.cc:33]: Could we add some comment on the size of the buffer (128 kB), like is it based on some estimation, or a guess that seems to work? (thinking a future reader that wonders why the size was set to 128 kB)

File: PerfTools/AllocMonitorPreload/src/memory_proxies.cc [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+  if (used != oldsize) {
+    reg.deallocCalled([]() {}, [oldsize]() { return oldsize; });
+    reg.allocCalled(
+        size, []() { return nullptr; }, [used](auto) { return used; });
+  }
+  return ret;
+}
+
+void* aligned_alloc(size_t alignment, size_t size) noexcept {
+  static auto original = get<decltype(&::aligned_alloc)>("aligned_alloc");
```
Comment (makortel) [inline PerfTools/AllocMonitorPreload/src/memory_proxies.cc:158]:
```suggestion
  static const auto original = get<decltype(&::aligned_alloc)>("aligned_alloc");
```
?

File: PerfTools/AllocMonitorPreload/src/memory_proxies.cc [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+
+  auto& reg = AllocMonitorRegistry::instance();
+  return reg.allocCalled(
+      size,
+      [alignment, size]() { return original(alignment, size); },
+      [](auto ret) { return malloc_usable_size(ret); });
+}
+
+void free(void* ptr) noexcept {
+  static auto original = get<decltype(&::free)>("free");
```
Comment (makortel) [inline PerfTools/AllocMonitorPreload/src/memory_proxies.cc:171]:
```suggestion
  static const auto original = get<decltype(&::free)>("free");
```
?

File: PerfTools/AllocMonitorPreload/src/memory_proxies.cc [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+//----------------------------------------------------------------
+//C++ memory functions
+
+#define CPP_MEM_OVERRIDE
+
+#if defined(CPP_MEM_OVERRIDE)
+#include <new>
+
+void* operator new(std::size_t size) {
+  static auto original = get<void* (*)(std::size_t)>("_Znwm");
```
Comment (makortel) [inline PerfTools/AllocMonitorPreload/src/memory_proxies.cc:194]:
```suggestion
  static const auto original = get<void* (*)(std::size_t)>("_Znwm");
```
and hereafter (I won't repeat the comment more)?

File: PerfTools/AllocMonitorPreload/src/memory_proxies.cc [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+      get<void (*)(void*, std::align_val_t, const std::nothrow_t&) noexcept>("_ZdlPvSt11align_val_tRKSt9nothrow_t");
+
+  if (not alloc_monitor_running_state()) {
+    original(ptr, al, tag);
+    return;
+  }
+  auto& reg = AllocMonitorRegistry::instance();
+  reg.deallocCalled([ptr, al, &tag]() { original(ptr, al, tag); }, [ptr]() { return malloc_usable_size(ptr); });
+}  //_ZdlPvSt11align_val_tRKSt9nothrow_t
+void operator delete[](void* ptr, std::align_val_t al, const std::nothrow_t& tag) noexcept {
```
Comment (makortel) [inline PerfTools/AllocMonitorPreload/src/memory_proxies.cc:407]:
```suggestion
}  //_ZdlPvSt11align_val_tRKSt9nothrow_t

void operator delete[](void* ptr, std::align_val_t al, const std::nothrow_t& tag) noexcept {
```

File: PerfTools/MaxMemoryPreload/README.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+## Reporting
+When the application ends, the monitor will report the following to standard error:
+
+- Total amount of bytes requested by all allocation calls during the job. Note that actual _used_ allocation can be greater than requested as the allocator may require additional memory be assigned.
+- The maximum amount of _used_ allocated memory that was in use at one time.
+- The amount of _used_ memory allocated during the job that has yet to be reclaimed by calling deallocation.
+- Number of calls made to allocation functions.
+- Number of calls made to deallocation functions.
+This service is multi-thread safe. Note that when run multi-threaded the maximum reported value will vary from job to job.
```
Comment (makortel) [inline PerfTools/MaxMemoryPreload/README.md:27]:
```suggestion
- Number of calls made to deallocation functions.

This service is multi-thread safe. Note that when run multi-threaded the maximum reported value will vary from job to job.
```

File: PerfTools/AllocMonitorPreload/src/memory_proxies.cc (hunk 1 above)
Comment (dan131riley) [reply PerfTools/AllocMonitorPreload/src/memory_proxies.cc:33]: The libstdc++ exception handling reserves an emergency buffer, which on ARM64 defaults to 72704 bytes.  That triggers a call to dlsym() to load malloc(), which allocates another 32 bytes for the error handler buffer.  So I took 72736 and rounded up to the next power of 2 to be on the safe side.

File: PerfTools/AllocMonitor/plugins/HistogrammingAllocMonitor.cc [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    void deallocCalled(size_t iActual) final {
+      auto& u = deallocUsed_[bin(iActual)];
+      u.fetch_add(1, std::memory_order_acq_rel);
+    }
+
+    void performanceReport() const {
+      auto log = edm::LogSystem("HistogrammingAllocMonitor");
+      log << "Memory Histogram"
+          << "\n   size                 allocated           deallocated"
+          << "\n                  requested      used          used";
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply PerfTools/AllocMonitor/plugins/HistogrammingAllocMonitor.cc]: the two words belong with `allocated`.

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
Comparisons show https://github.com/cms-sw/cmssw/issues/39803

**Comment by makortel** [comment]
Comparisons show https://github.com/cms-sw/cmssw/issues/39803 again, although to lesser extent. This time both baseline and PR tests were run on `Intel(R) Xeon(R) CPU E5-2683 v4`.

**Comment by makortel** [comment]
+core

**Comment by rappoccio** [comment]
+1

**Comment by smuzaffar** [comment]
@Dr15Jones , should this work for CentOS7 ? The unit tests in IBs are failing for slc7 https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc11/CMSSW_13_3_X_2023-09-04-1100/unitTestLogs/PerfTools/AllocMonitor#/

**Comment by Dr15Jones** [comment]
@smuzaffar the expectation was it should work for CentOS7.

**Comment by dan131riley** [comment]
The failures are all timeouts, so guess it’s deadlocking somewhere?

**Comment by Dr15Jones** [comment]
#42724 fixes the problem. sl7 needs Dan's fix as well.

**Pull request merged**